packages/cli/lib/base-generator.js

```diff
@@ -28,11 +31,218 @@ module.exports = class BaseGenerator extends Generator {
    * Subclasses can extend _setupGenerator() to set up the generator
    */
   _setupGenerator() {
+    this.option('config', {
+      type: String,
+      alias: 'c',
+      description: 'JSON file to configure options',
+    });
+
+    this.option('skip-optional-prompts', {
+      type: Boolean,
+      alias: 'b',
+      description:
+        'Run the generator in the mode without prompting optional questions',
+    });
+
     this.artifactInfo = this.artifactInfo || {
       rootDir: 'src',
     };
   }
 
+  /**
+   * Read a json document from stdin
+   */
+  _readJSONFromStdin() {
+    const rl = readline.createInterface({
+      input: process.stdin,
+    });
+
+    const lines = [];
+    if (process.stdin.isTTY) {
+      this.log(
+        chalk.green(
+          'Please type in a json object line by line ' +
```

Comment:
Uh oh, do you really expect people to type JSON by hand? What if they make a typo, e.g. forget to wrap a key in double quotes - how are they going to fix the type? Will they have to re-type everything again?

In my opinion, adding readline support is overkill. The JSON mode is intended for machines, humans should use interactive prompt mode.

Reply:
+1

Reply:
@raymondfeng PTAL ☝️

Reply:
I'm thinking about `copy-paste` of json objects into `stdin`.

Reply:
Please also note the `readline` support is needed for unix pipe or redirection, for example:

- `cat config.json | lb4 app --config stdin` 
- `lb4 app --config stdin < config.json`

Reply:
> I'm thinking about copy-paste of json objects into stdin.

I usually do `pbpaste | lb4 app --config stdin` on my Mac. IDK if there are similar CLI tools for other platforms though.

On UNIX, you can also use the `<<EOF` syntax you have shown above.

```
lb4 app --config stdin <<EOF
(paste the content here)
EOF
```

IMO, we should not be duplicating the feature of Unix shell in our tooling. Let's keep our code small and simple please.

> Please also note the readline support is needed for unix pipe or redirection.

I don't understand why is readline needed for unix pipes, could you please explain?

Reply:
The generator needs to read from `stdin` - which is the source of input from unix pipes. The `readline` module is used to enable that.

Reply:
The other option is the use the following code:
```js
function readBufferFromStdin() {
	const stdin = process.stdin;

	const data = [];
	let len = 0;

	return new Promise((resolve, reject) => {

		stdin.on('readable', () => {
			let chunk;

			while ((chunk = stdin.read())) {
				data.push(chunk);
				len += chunk.length;
			}
		});

		stdin.on('end', () => {
			resolve(Buffer.concat(data, len));
		});

		stdin.on('error', err => {
			reject(err);
		});
	});
};
```

Reply:
> Well, that's just a different way of implementing readJSONFromStdin. Is leveraging readline to read from stdin a bad idea?

I suppose using `readline` is not a bad idea per se. My main objection is against the complexity of the current implementation of `_readJSONFromStdin`. To me, the currently proposed code requires maintainers to know a lot about readline, TTY, SIGINT, etc. and the way how all things are wired together.

Compare it with my proposed solution that's just 5 lines of relatively straightforward code.

```js
return new Promise((resolve, reject) => {
  const concatStream = concat(resolve);
  src.on('error', reject);
  src.pipe(concatStream);
});
```

Your second example, using ReadableStream directly, looks better to me than the original readline-based one. But why to re-implement reading a stream into a buffer when `concat-stream` module already does it well? Dealing with streams is tricky and it's very easy to introduce subtle bugs that are difficult to spot by casual readers. 

Not a big deal though. If you prefer a hand-written `readBufferFromStdin` then I am ok with that too.